### PR TITLE
Partition Encryption during Installation

### PR DESCRIPTION
**Pull request by mdhorn**
Fixes Issue: #10 
Fixes Issue: #101 

Changes proposed in this pull request:
- Add support for partition encryption
- Add support for partition labels
- Fixes bug in vendored files
- Add new test to ensure the integrity of vendored files

### CODE REVIEW

**Comment by mdhorn** [comment]
@dorileo I'm working on improving test coverage, but I'd like you to start reviewing these two new features and provide feedback. I'll also have to update the bundles for encryption once the bundle updates make it into the next Clear Linux OS release.

**Comment by dorileo** [comment]
@mdhorn can you break the first 3 patches into another PR so we can merge it right away?

I mean these 3 patches:

Fix missing vendor files a26116e
Add check to ensure vendor file status 88a0fac
New target for adding new Vendor packages b77d2a9

File: controller/controller.go [hunk 1]
```diff
@@ -270,6 +300,19 @@ func Install(rootDir string, model *model.SystemInstall, options args.Args) erro
...
 
+	if encryptedUsed {
+		model.AddBundle(storage.RequiredBundle1)
+		model.AddBundle(storage.RequiredBundle2)
+	}
+
+	msg := fmt.Sprintf("Writing mount files")
+	prg = progress.NewLoop(msg)
+	log.Info(msg)
+	if err = storage.GenerateTabFiles(rootDir, model.TargetMedias); err != nil {
```
Comment (dorileo) [inline controller/controller.go]: so we really need a fstab file then?

File: tui/popup_passphrase.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+	}
+
+	dialog.cancelButton.OnClick(func(ev clui.Event) {
+		dialog.Confirmed = false
+		dialog.Close()
+	})
+
+	dialog.confirmButton.OnClick(func(ev clui.Event) {
+		dialog.Confirmed = true
+		modelSI.CryptPass = dialog.passphraseEdit.Title()
```
Comment (dorileo) [inline tui/popup_passphrase.go]: so we're dealing with plain passphrase?

File: storage/storage.go [hunk 3]
```diff
@@ -495,6 +540,35 @@ func listBlockDevices(userDefined []*BlockDevice) ([]*BlockDevice, error) {
...
+func UpdateBlockDevices(medias []*BlockDevice) error {
+
+	bds, err := listBlockDevices(nil)
+	if err != nil {
+		return err
+	}
+
+	// Loop though all used media devices and update
+	// the labels and uuid
+	for _, media := range medias {
```
Comment (dorileo) [inline storage/storage.go]: can we make it recursive and check the device no matter the level? :)

File: storage/encrypt.go [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+// EncryptionRequiresPassphrase checks all partition to see if encryption was enabled
+func (bd *BlockDevice) EncryptionRequiresPassphrase() bool {
+	enabled := (bd.Type == BlockDeviceTypeCrypt && bd.FsType != "swap")
+
+	for _, ch := range bd.Children {
+		if len(ch.Children) > 0 {
+			enabled = enabled || ch.EncryptionRequiresPassphrase()
+		} else {
+			enabled = enabled || (ch.Type == BlockDeviceTypeCrypt &&
+				ch.FsType != "swap")
```
Comment (dorileo) [inline storage/encrypt.go]: can we replace all these ```ch.FsType != "swap"``` for a function call ?

**Comment by dorileo** [comment]
@mdhorn the patches look consistent, had just minor comments. I'll take some time to test it yet today. Thanks.

**Comment by mdhorn** [comment]
> @mdhorn can you break the first 3 patches into another PR so we can merge it right away?
> 
> I mean these 3 patches:
> 
> Fix missing vendor files [a26116e](https://github.com/clearlinux/clr-installer/commit/a26116ec493dccbddbb3f68ebfdd82482579a325)
> Add check to ensure vendor file status [88a0fac](https://github.com/clearlinux/clr-installer/commit/88a0fac0db181fa1090e8d1fd876d60bf74947d0)
> New target for adding new Vendor packages [b77d2a9](https://github.com/clearlinux/clr-installer/commit/b77d2a98c98f1e3ffcf23f2bed8af0107fdc9ff6)

Yep, can do.

File: controller/controller.go (hunk 1 above)
Comment (mdhorn) [reply controller/controller.go]: Yes, we really need the /etc/fstab and /etc/crypttab if we have encrypted swap or a non-root encrypted partition. We also need just /etc/fstab if someone creates a non-standard partition; aka not /, boot, /home, or /srv

File: tui/popup_passphrase.go (hunk 2 above)
Comment (mdhorn) [reply tui/popup_passphrase.go]: Yes, there is no hashed version of the passphrase. But, we DO NOT store the passphrase in the YAML file nor do we ever expose the plain text passphrase on a command line.

File: storage/encrypt.go (hunk 4 above)
Comment (mdhorn) [reply storage/encrypt.go]: there are only three occurrences, one each in three different source files. You think making that logic into a function call makes it more readable?

File: controller/controller.go (hunk 1 above)
Comment (dorileo) [reply controller/controller.go]: ok

File: tui/popup_passphrase.go (hunk 2 above)
Comment (dorileo) [reply tui/popup_passphrase.go]: ok

File: storage/encrypt.go (hunk 4 above)
Comment (dorileo) [reply storage/encrypt.go]: makes it more expandable

**Comment by dorileo** [comment]
@mdhorn  If I have previously installed a system with a encrypted partition and try to install with the same media the installer will break.

**How to reproduce:**
- perform an install with a root partition encrypted
    - go to media configuration and select auto partition
    - select root partition and mark as encrypted
    - install
- try to install to the same media
  - go to media configuration and select auto partition

**Comment by dorileo** [comment]
@mdhorn failed to boot on a bare metal system (it hangs after typing the passphrase). It worked fine with qemu.

**Comment by mdhorn** [comment]
> @mdhorn failed to boot on a bare metal system (it hangs after typing the passphrase). It worked fine with qemu.

@dorileo I'm seeing the same hang with the old ister_gui installer when attempting to use encrypted root partition. I'm beginning to think there is a problem with the latest cryptsetup 2.0.5 which was recently released.

**Comment by neben** [comment]
I'd be happy to test this on a NUC8i7BEH if you provide me with an installer image.

File: storage/encrypt.go (hunk 4 above)
Comment (mdhorn) [reply storage/encrypt.go]: @dorileo Updated in latest push

**Comment by mdhorn** [comment]
@neben The images are a bit large for me to post, but you can create one for yourself.

```
git clone https://github.com/mdhorn/clr-installer.git
cd clr-installer
git checkout encryption
curl -O https://gist.githubusercontent.com/mdhorn/460b686efe102b6ab458b98a70c95bd2/raw/a3d5f33d4caa0c541929851b449db8337ca6fda7/live-encrypted.yaml
make
sudo .gopath/bin/clr-installer -c live-encrypted.yaml
```

**Comment by mdhorn** [comment]
@dorileo I was not able to reproduce the issue noted about reusing media which was previously encrypted.
Can you retest with the latest update and latest version of Clear Linux OS which has the updated bundles?
Also, please verify your code review concerns where address in this latest update.

**Comment by mrkz** [comment]
how can I grab output/log from the installer if it crashed on the middle of a new install?

(I got `clear-26500-live-desktop-beta.img`, followed the results from the [previous comment](https://github.com/clearlinux/clr-installer/pull/169#issuecomment-441860961)) and copied the new compiled clr-installer into the usb-image.

I'm trying to install with encrypted root on a previously root-encrypted device fwiw.

**Comment by dorileo** [comment]
@mrkz there's a /usr/clr-installer.log file.

**Comment by mrkz** [comment]
@dorileo @mdhorn this is the output from the log I've got:
```
2018/11/27 09:36:27 [INF] clr-installer: 1.0.0-71-g002fbc5, built on 2018-11-27_15:25:04_UTC
2018/11/27 09:36:27 [WRN] PrivateIP: Could not determine network location: lookup clr.telemetry.intel.com on [2001:4860:4860::8888]:53: dial udp [2001:4860:4860::8888]:53: connect: network is unreachable
2018/11/27 09:43:23 [INF] Testing connectivity
2018/11/27 09:45:19 [INF] Running pre-install hooks
2018/11/27 09:45:20 [INF] Querying Clear Linux version
2018/11/27 09:45:20 [INF] Writing partition table to: sda
2018/11/27 09:45:22 [INF] Adjusting filesystem configurations
2018/11/27 09:45:31 [INF] Writing vfat file system to sda1 '/boot'
2018/11/27 09:45:33 [INF] Writing swap file system to sda2
2018/11/27 09:45:35 [INF] Mapping sda3 partition to an encrypted partition
2018/11/27 09:45:43 [INF] Writing ext4 file system to sda3 '/'
2018/11/27 09:45:47 [INF] Writing ext4 file system to sda4 '/home'
2018/11/27 09:45:47 [ERR] 

Error Trace:
storage.(*BlockDevice).MakeFs()
     .gopath/src/github.com/clearlinux/ops.go:57
controller.Install()
     .gopath/src/github.com/clearlinux/controller.go:235
 Trying to run MakeFs() against a disk, partition required
```

**Comment by mrkz** [comment]
> @mrkz there's a /usr/clr-installer.log file.

Actually, found them in `/root/clr-installer.log` and `/home/clr-installer.log`, so I think it places the log in $HOME.

**Comment by mdhorn** [comment]
> > @mrkz there's a /usr/clr-installer.log file.
> 
> Actually, found them in `/root/clr-installer.log` and `/home/clr-installer.log`, so I think it places the log in $HOME.

Yes, the log defaults to $HOME, you can override this with       `--log-file <string>`

**Comment by neben** [comment]
> @neben The images are a bit large for me to post, but you can create one for yourself.

I've added an encrypted swap partition to your gist (just to test), and it boots up fine on my NUC from a usb stick. The timeout issue seems fixed, too.

The image creation script finishes successfully, however, in the log file there's the following error (image is created fine, though):
```
[...]
2018/11/27 16:14:21 [ERR] Failed to log Telemetry success record
2018/11/27 16:14:21 [INF] Skipping archiving of Installation results
2018/11/27 16:14:23 [WRN] Failed to stop image Telemetry server
2018/11/27 16:14:23 [WRN] Failure accessing a path "/var/spool/telemetry": lstat /var/spool/telemetry: no such file or directory

2018/11/27 16:14:23 [ERR] 

Error Trace:
controller.saveInstallResults()
     .gopath/src/github.com/clearlinux/controller.go:709
controller.Install()
     .gopath/src/github.com/clearlinux/controller.go:388
 Failed to stop image Telemetry server
2018/11/27 16:14:23 [INF] Umounting rootDir: /tmp/install-360651031
2018/11/27 16:14:23 [INF] Removing rootDir: /tmp/install-360651031
```

Do you have an estimate when this will appear in the official installer images?

**Comment by mdhorn** [comment]
> @dorileo @mdhorn this is the output from the log I've got:
> 
> ```
> 2018/11/27 09:36:27 [INF] clr-installer: 1.0.0-71-g002fbc5, built on 2018-11-27_15:25:04_UTC
> 2018/11/27 09:36:27 [WRN] PrivateIP: Could not determine network location: lookup clr.telemetry.domain.com on [2001:4860:4860::8888]:53: dial udp [2001:4860:4860::8888]:53: connect: network is unreachable
> 2018/11/27 09:43:23 [INF] Testing connectivity
> 2018/11/27 09:45:19 [INF] Running pre-install hooks
> 2018/11/27 09:45:20 [INF] Querying Clear Linux version
> 2018/11/27 09:45:20 [INF] Writing partition table to: sda
> 2018/11/27 09:45:22 [INF] Adjusting filesystem configurations
> 2018/11/27 09:45:31 [INF] Writing vfat file system to sda1 '/boot'
> 2018/11/27 09:45:33 [INF] Writing swap file system to sda2
> 2018/11/27 09:45:35 [INF] Mapping sda3 partition to an encrypted partition
> 2018/11/27 09:45:43 [INF] Writing ext4 file system to sda3 '/'
> 2018/11/27 09:45:47 [INF] Writing ext4 file system to sda4 '/home'
> 2018/11/27 09:45:47 [ERR] 
> 
> Error Trace:
> storage.(*BlockDevice).MakeFs()
>      .gopath/src/github.com/clearlinux/ops.go:57
> controller.Install()
>      .gopath/src/github.com/clearlinux/controller.go:235
>  Trying to run MakeFs() against a disk, partition required
> ```

@mrkz I'm still have an issue reproducing this, can you share your YAML file you were using when the crash occurred? Also, if you can reproduce this using `-l 4` to increase the logging, that would be useful.

**Comment by mrkz** [comment]
>  @mrkz I'm still have an issue reproducing this, can you share your YAML file you were using when the crash occurred? Also, if you can reproduce this using -l 4 to increase the logging, that would be useful.

I did not use any yaml file, I launched `/usr/bin/clr-installer` as root in a terminal window and filled all the options, then I started the installed and failed.

Interestingly enough, now that I launched the installer for the third time (now using `-l 4` option) did not fail.

I'll try few more times to reproduce back again with `-l 4` option to get more info.

**Comment by mrkz** [comment]
I tried with @mdhorn new pushes and I can confirm 428c328 looks neat from the main page, I was wondering if we could specify somewhere in the main page what the `*` means (probably in the very same disk section a parenthesis saying something like `( * : encrypted partition)` )

>I'll try few more times to reproduce back again with -l 4 option to get more info.

So far I haven't been able to reproduce the failure back again

**Comment by mdhorn** [comment]
> I tried with @mdhorn new pushes and I can confirm [428c328](https://github.com/clearlinux/clr-installer/commit/428c328ce56d709d3b59e0942e597c1c33e4d91f) looks neat from the main page, I was wondering if we could specify somewhere in the main page what the `*` means (probably in the very same disk section a parenthesis saying something like `( * : encrypted partition)` )
> 
> > I'll try few more times to reproduce back again with -l 4 option to get more info.
> 
> So far I haven't been able to reproduce the failure back again

Thanks for testing. If I can not get you or Leandro to reproduce by the end of the day, and Leandro finishes his code review, I will likely merge end of the day.

I thought about putting a 'key for the *', but two problems; 1) there is limited screen space of 80x24 which we need to maintain as the TUI was intended for non-graphical environments, and 2) we were requested by documentation and marketing to put the '*' after "Clear Linux@ OS Installer" on the top of the page so the '*' is used in two ways.

**Comment by mrkz** [comment]
I ran into 2 issues after installing:

 1) looks like the entropy is rather low when booting into the encrypted-root device, it took me several seconds (probably around a minute after typing the passphrase) to get it to the gnome-initial-setup screen (didn't create any user on the installer). At some point I thought the initramfs got stuck somewhere and failed.

photo_2018-11-28_12-53-31

2) I cannot boot directly into the new Clear Linux* OS installation, not sure it's related to the encrypted PR, so I created a new issue [here #174 ](https://github.com/clearlinux/clr-installer/issues/174)

This is the disk scheme I've created for the installation:

```
# lsblk 
NAME                                          MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
sda                                             8:0    0 447.1G  0 disk  
├─sda1                                          8:1    0   143M  0 part  /boot
├─sda2                                          8:2    0  15.3G  0 part  
├─sda3                                          8:3    0 238.4G  0 part  
│ └─luks-0a0f2cfe-5f96-4185-8bad-d1dd7c141cb3 252:0    0 238.4G  0 crypt /
└─sda4                                          8:4    0 172.6G  0 part  
  └─home                                      252:1    0 172.6G  0 crypt /home
```

(btw, kudos on handling the encryption passphrase for both partitions `/` and `/home`) :tada:

**Comment by dorileo** [comment]
@mrkz can you test the same scenario with ister and see if it also takes that long to boot? and can you test with only root partition encrypted and see if you see the same behavior?

**Comment by mrkz** [comment]
> @mrkz can you test the same scenario with ister and see if it also takes that long to boot? and can you test with only root partition encrypted and see if you see the same behavior?

I tried with ister and didn't take long to boot after typing passphrase
I tried with clr-installer (only root encryption) and didn't take long either

I will re-try with both `/home` and `/` encrypted.

**Comment by mdhorn** [comment]
> > @mrkz can you test the same scenario with ister and see if it also takes that long to boot? and can you test with only root partition encrypted and see if you see the same behavior?
> 
> I tried with ister and didn't take long to boot after typing passphrase
> I tried with clr-installer (only root encryption) and didn't take long either
> 
> I will re-try with both `/home` and `/` encrypted.

@mrkz Do you believe the boot time is equal? Both old and new installers are using the same clr-init to handle the encryption for file systems.  I've rebased with the other latest fixes, I think we are ready to merge this.

**Comment by mrkz** [comment]
> @mrkz Do you believe the boot time is equal? Both old and new installers are using the same clr-init to handle the encryption for file systems. I've rebased with the other latest fixes, I think we are ready to merge this.

Yes apparently they take the same time (didn't run any timer or anything, but visually takes the same time to boot).

**Pull request merged**